crates/sui-core/src/overload_monitor.rs

```diff
@@ -0,0 +1,292 @@
+        self.load_shedding_percentage.store(0, Ordering::Relaxed);
+    }
+}
+
+// Monitors the overload signals in `authority_state` periodically, and updates its `overload_info`
+// when the signals indicates overload.
+pub async fn overload_monitor(
+    authority_state: Weak<AuthorityState>,
+    config: OverloadThresholdConfig,
+) {
+    info!("Starting system overload monitor.");
+
+    loop {
+        let authority_exist = check_authority_overload(&authority_state, &config);
+        if !authority_exist {
+            // `authority_state` doesn't exist anymore. Quit overload monitor.
+            break;
+        }
+        sleep(config.overload_monitor_interval).await;
+    }
+
+    info!("Shut down system overload monitor.");
+}
+
+// Checks authority overload signals, and updates authority's `overload_info`.
+// Returns whether the authority state exists.
+fn check_authority_overload(
+    authority_state: &Weak<AuthorityState>,
+    config: &OverloadThresholdConfig,
+) -> bool {
+    let authority_arc = authority_state.upgrade();
+    if authority_arc.is_none() {
+        // `authority_state` doesn't exist anymore.
+        return false;
+    }
+
+    let authority = authority_arc.unwrap();
+    let queueing_latency = authority
+        .metrics
+        .execution_queueing_latency
+        .latency()
+        .unwrap_or_default();
+    let txn_ready_rate = authority.metrics.txn_ready_rate_tracker.lock().rate();
+    let execution_rate = authority.metrics.execution_rate_tracker.lock().rate();
+
+    let (is_overload, load_shedding_percentage) =
+        check_overload_signals(config, queueing_latency, txn_ready_rate, execution_rate);
+    if is_overload {
+        authority
+            .overload_info
+            .set_overload(load_shedding_percentage);
+    } else {
+        authority.overload_info.clear_overload();
+    }
+
+    authority
+        .metrics
+        .authority_overload_status
+        .set(is_overload as i64);
+    authority
+        .metrics
+        .authority_load_shedding_percentage
+        .set(load_shedding_percentage as i64);
+    true
+}
+
+// Calculates the percentage of transactions to drop in order to reduce execution queue.
+// Returns the integer percentage between 0 and 100.
+fn calculate_load_shedding_percentage(txn_ready_rate: f64, execution_rate: f64) -> u32 {
+    // When transaction ready rate is practically 0, we aren't adding more load to the
+    // execution driver, so no shedding.
+    // TODO: consensus handler or transaction manager can also be overloaded.
+    if txn_ready_rate < 1e-10 {
+        return 0;
+    }
+
+    // Deflate the execution rate to account for the case that execution_rate is close to
+    // txn_ready_rate.
+    if execution_rate * 0.9 > txn_ready_rate {
+        return 0;
+    }
+
+    // In order to maintain execution queue length, we need to drop at least (1 - executionRate / readyRate).
+    // TO reduce the queue length, here we add 10% more transactions to drop.
```

Comment:
I'm not sure I understand the match here, especially the `0.9` and `0.1` constants. So, suppose we are 2x overloaded, e.g. our execution rate is 5 and our ready rate is 10.

Then we get (1.0 - 5 * 0.9 / 10) + 0.1 which is 0.65, so we are dropping 30% more transactions than we need to in order to make the incoming and outgoing rates equal.

Reply:
Yeah, I'm also not 100% confident with these const. Your calculation is mostly right.

0.9 is account for the case where execution rate ~= ready rate. In a steady state, these two rates are probably going to oscillate. One moment, it's (1997, 2001), and another moment, it's (2003, 1998) . So here I want to make sure that when we say ready rate is higher, is indeed higher. Ideally, the 0.9 should be replaced with oscillation rate in a steady state.

For the 0.1, it's for the case where we want the queueing delay to drop. In your example, simply dropping 50% of requests only make enqueue and dequeue rate to match. So we need a bit more to make it to drop. I also thought about maybe use *1.1 instead of +0.1. I don't know which one is better tbo.

I think these two are those parameters that it's hard to find the best value. So I use conservative numbers to start with. My plan is to set it here and tweak it when I test it in private testnet (hopefully one or two rounds of test will show whether they are ok or not).